tools/cndump/fatal.h

```diff
@@ -0,0 +1,18 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#ifndef CNDUMP_FATAL_H
+#define CNDUMP_FATAL_H
+
+#include <hse/error/merr.h>
+#include <hse_util/compiler.h>
+
+void
+fatal(const char *who, merr_t err);
+
+void HSE_PRINTF(1, 2)
+syntax(const char *fmt, ...);
```

Comment:
Were you going to move the decl to the end?

Can you add support for the `noreturn` attribute to compiler.h, and then use it here?

https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html

Also we have a million of these syntax functions in the code. They should be consolidated at some point.

Reply:
It was this:
```
HSE_PRINTF(1, 2)
void syntax(const char *fmt, ...);
```
You said "end of decl".  So I tried this, but it didn't compile:
```
void syntax(const char *fmt, ...) HSE_PRINTF(1, 2);
```

I figured maybe you meant end of type, which worked.

Perhaps I fat-fingered my attempt to put it and end of function...

Reply:
Yeah, works for me.  Now if you left off the (1,2) you will get an explosion of compiler error messages...

Reply:
Regarding "noreturn", why not use C11 keyword `_Noreturn` instead of the GNU function attribute?

Reply:
There is this:

> An attribute specifier list may appear immediately before the comma, = or semicolon terminating the declaration of an identifier other than a function definition

Odd that the examples are all over the map.

Reply:
Generally, for function attributes, I put the attribute at the front of a declaration and at the end of a definition.  Some attributes require this, others don't, but I think the general rule works most of the time...
When in front of a declaration it can sometimes matter precisely where in front, I like to make it the very first thing on a line by itself when possible.  IBM used to have a good page about this stuff, since they make a compiler (xlc) that needs to be able to consume GNUisms...

Reply:
For funciton decls/prototypes: `<type> <name> ( <params> ) <attrs>;`
For function defs/impelementaions:  `<type> <attrs> <name> ( <params> ) { <body> };`

Reply:
clangd will warn you if your attribute is in the wrong place. I was just suggesting moving it for consistency.

As far as `_Noreturn`, use that. I forget about it 😄

Reply:
Turns out `_Noreturn` is being deprecated in C23.  The future proof way is to use `<stdnoreturn.h>` which is available with `gcc -std=gnu11`.

Relevant patch in this PR:
```
--- a/lib/util/include/hse_util/compiler.h
+++ b/lib/util/include/hse_util/compiler.h
@@ -115,6 +115,13 @@
+#if defined(__has_include) && __has_include(<stdnoreturn.h>)
+#include <stdnoreturn.h>
+#define HSE_NORETURN noreturn
+#else
+#define HSE_NORETURN
+#endif
+
```